lib/ex_unit/lib/ex_unit/manifest.ex

```diff
@@ -16,7 +17,25 @@ defmodule ExUnit.Manifest do
     []
   end
 
+  @spec to_last_run_status_index(t) :: last_run_status_index
+  def to_last_run_status_index(manifest) do
+    Map.new(manifest, fn {test_id, entry(last_run_status: status)} ->
+      {test_id, status}
+    end)
+  end
+
+  @spec get_files_with_failures(t) :: MapSet.t(Path.t())
+  def get_files_with_failures(entries) do
+    entries
+    |> Stream.filter(fn {_, entry(last_run_status: status)} -> status == :failed end)
+    |> MapSet.new(fn {_, entry(file: file)} -> file end)
```

Comment:
I think the fastest implementation of this function would be:

    for  {_, entry(last_run_status: :failed, file: file)} <- entries, do: file, uniq: true

We can do everything in one pass and we can return a simple list back, which is what we end-up traversing later on anyway.

Reply:
I think it's important that we return a set.  For one, conceptually, a set is the right data structure since we want to return a collection of unique files and we don't care about the order.  More importantly, we want a constant-time [membership check](https://github.com/elixir-lang/elixir/pull/7373/files#diff-3b89dff3429a39440b077c60dd5cd932R325) when we use this later in mix.  Otherwise, the `filter_test_files_using_manifest/3` function in mix is going to be an `O(n * m)` operation (where n == number of test files and m == number of test files with failures).

> We can do everything in one pass

With the `Stream.filter`, we're only traversing the list once, anyway, right?

On a side note, I didn't know that `for` comprehensions support a `uniq` option.  TIL!